core/node/consensus/src/storage/vm_reader.rs

```diff
@@ -0,0 +1,341 @@
+            })
+        }
+
+        validators
+    }
+
+    pub async fn read_attester_committee(
+        &mut self,
+        ctx: &Ctx,
+        block_id: BlockId,
+    ) -> Vec<CommitteeAttester> {
+        let mut attesters = vec![];
+        let num_committee_attesters = self.read_num_committee_attesters(ctx, block_id).await;
+        for i in 0..num_committee_attesters {
+            let committee_attester = self.read_committee_attester(ctx, block_id, i).await;
+            let attester = self
+                .read_attester(ctx, block_id, committee_attester.1)
+                .await;
+
+            attesters.push(CommitteeAttester {
+                node_owner: committee_attester.1,
+                weight: committee_attester.0,
+                pub_key: attester.1,
+            })
+        }
+        attesters
+    }
+
+    async fn read_num_committee_validators(&mut self, ctx: &Ctx, block_id: BlockId) -> usize {
+        let func = self
+            .validator_registry_contract
+            .function("numCommitteeValidators")
+            .unwrap()
+            .clone();
+
+        let tx = self.gen_l2_call_tx(
+            self.validator_registry_address.unwrap(),
+            func.short_signature().to_vec(),
+        );
+
+        let res = self.eth_call(ctx, block_id, tx).await;
+
+        func.decode_output(&res).unwrap()[0]
+            .clone()
+            .into_uint()
+            .unwrap()
+            .as_usize()
+    }
+
+    async fn read_num_committee_attesters(&mut self, ctx: &Ctx, block_id: BlockId) -> usize {
+        let func = self
+            .attester_registry_contract
+            .function("numCommitteeAttesters")
+            .unwrap()
+            .clone();
+        let tx = self.gen_l2_call_tx(
+            self.attester_registry_address.unwrap(),
+            func.short_signature().to_vec(),
+        );
+
+        let res = self.eth_call(ctx, block_id, tx).await;
+        func.decode_output(&res).unwrap()[0]
+            .clone()
+            .into_uint()
+            .unwrap()
+            .as_usize()
+    }
+
+    async fn read_validator(
+        &mut self,
+        ctx: &Ctx,
+        block_id: BlockId,
+        node_owner: zksync_types::ethabi::Address,
+    ) -> (usize, Vec<u8>, Vec<u8>, bool) {
+        let func = self
+            .validator_registry_contract
+            .function("validators")
+            .unwrap()
+            .clone();
+        let tx = self.gen_l2_call_tx(
+            self.validator_registry_address.unwrap(),
+            func.encode_input(&[Token::Address(node_owner)]).unwrap(),
+        );
+        let res = self.eth_call(ctx, block_id, tx).await;
+        let tokens = func.decode_output(&res).unwrap();
+        (
+            tokens[0].clone().into_uint().unwrap().as_usize(),
+            tokens[1].clone().into_bytes().unwrap(),
+            tokens[2].clone().into_bytes().unwrap(),
+            tokens[3].clone().into_bool().unwrap(),
+        )
+    }
```

Comment:
Probably it would be better to implement `zksync_types::web3::contract::Tokenize` on all the required structures instead of mixing parsing with the business logic.

Reply:
Changes here: https://github.com/matter-labs/zksync-era/pull/2416/commits/477ea5ec1acc95cede12a576d8d735b1ba1978dd